trackpy/motion.py

```diff
@@ -35,26 +36,120 @@ def msd(traj, mpp, fps, max_lagtime=100, detail=False, pos_columns=['x', 'y']):
+
+
+def _msd_iter(pos, lagtimes):
+    for lt in lagtimes:
+        diff = pos[lt:] - pos[:-lt]
+        yield np.concatenate((np.nanmean(diff, axis=0),
+                              np.nanmean(diff**2, axis=0)))
+
+
+def _msd_gaps(traj, mpp, fps, max_lagtime=100, detail=False, pos_columns=None):
+    """Compute the mean displacement and mean squared displacement of one
+    trajectory over a range of time intervals."""
+    if pos_columns is None:
+        pos_columns = ['x', 'y']
+    result_columns = ['<{}>'.format(p) for p in pos_columns] + \
+                     ['<{}^2>'.format(p) for p in pos_columns]
+
     # Reindex with consecutive frames, placing NaNs in the gaps.
+    pos = traj.set_index('frame')[pos_columns] * mpp
     pos = pos.reindex(np.arange(pos.index[0], 1 + pos.index[-1]))
-    max_lagtime = min(max_lagtime, len(t))  # checking to be safe
-    lagtimes = 1 + np.arange(max_lagtime)
-    disp = pd.concat([pos.sub(pos.shift(lt)) for lt in lagtimes],
-                     keys=lagtimes, names=['lagt', 'frames'])
-    results = mpp*disp.mean(level=0)
-    results.columns = ['<{}>'.format(p) for p in pos_columns]
-    results[['<{}^2>'.format(p) for p in pos_columns]] = mpp**2*(disp**2).mean(level=0)
-    results['msd'] = mpp**2*(disp**2).mean(level=0).sum(1) # <r^2>
-    # Estimated statistically independent measurements = 2N/t
+
+    max_lagtime = min(max_lagtime, len(pos) - 1)  # checking to be safe
+
+    lagtimes = np.arange(1, max_lagtime + 1)
+
+    result = pd.DataFrame(_msd_iter(pos.values, lagtimes),
+                          columns=result_columns, index=lagtimes)
+    result['msd'] = result[result_columns[-len(pos_columns):]].sum(1)
     if detail:
-        results['N'] = 2*disp.iloc[:,0].count(level=0).div(Series(lagtimes))
-    results['lagt'] = results.index.values/fps
-    return results[:-1]
+        # effective number of measurements
+        # approximately corrected with number of gaps
+        result['N'] = _msd_N(len(pos), lagtimes) * len(traj) / len(pos)
+    result['lagt'] = result.index.values/float(fps)
+    result.index.name = 'lagt'
+    return result
+
 
+def _msd_fft(traj, mpp, fps, max_lagtime=100, detail=False, pos_columns=None):
+    """Compute the mean displacement and mean squared displacement of one
+    trajectory over a range of time intervals using FFT transformation.
+
+    The original Python implementation comes from a SO answer :
+    http://stackoverflow.com/questions/34222272/computing-mean-square-displacement-using-python-and-fft#34222273.
+    The algorithm is described in this paper : http://dx.doi.org/10.1051/sfn/201112010.
+    """
+    if pos_columns is None:
+        pos_columns = ['x', 'y']
+    result_columns = ['<{}>'.format(p) for p in pos_columns] + \
+                     ['<{}^2>'.format(p) for p in pos_columns]
 
-def imsd(traj, mpp, fps, max_lagtime=100, statistic='msd', pos_columns=['x', 'y']):
+    r = traj[pos_columns].values * mpp
```

Comment:
The DataFrame -> numpy conversion is currently necessary. It's also done in `_msd_gaps`. For instance at line 127 the code would fail if you don't do the conversion. (dataframes are aligned on index values)

We might try to do a full implementation on dataframes but it could be that the higher efficiency of numpy compensates for the copy on this line.